### PR TITLE
[4.0] Improve assets handling

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue https://github.com/joomla/joomla-cms/pull/19637 .

### Summary of Changes
This improves the way assets are handled. 
So it has a logic of first loonking for the minified and then the non minified version if in normal mode and if in debugging mode it reverses the order. 

In plain english even if you do `JHtml::_('script', 'my-awesome-script.js');` or 
`JHtml::_('script', 'my-awesome-script.min.js');` it doesn't matter! Joomla will pick the right one for you! 

### Testing Instructions
Apply patch, rename /media/system/js/core.min.js to  /media/system/js/x-core.min.js
core.js is loaded, revert the change
Enable debug mode
rename /media/system/js/core.js to  /media/system/js/x-core.js
core.min.js is loaded, revert the change
Test is successful!

### Documentation Changes Required
No

There is a change from enabling browser dependant script to be searched automatically to false, this is a performance improvement (checking for less files), but the functionality remains there, it just needs the right attribute to be passed in the JHtml::_()

This are the php changes to remove the minified version of assets, the rest will come when/if the rest of the affect scripts gets updated

### CODE REVIEW

File: libraries/src/HTML/HTMLHelper.php [hunk 1]
```diff
@@ -404,230 +405,203 @@ protected static function getMd5Version($path)
...
-							// Divide the file extracting the extension as the first part before /
-							list($extension, $file) = explode('/', $file, 2);
+							$includes[] = $found;
 
-							// If the file yet contains any /: it can be a plugin
+							break;
+						}
+						else
+						{
+							// If the file contains any /: it can be in an media extension subfolder
```
Comment (brianteeman) [inline libraries/src/HTML/HTMLHelper.php]: just to prove i read these 
an ==> a

File: libraries/src/HTML/HTMLHelper.php [hunk 2]
```diff
@@ -404,230 +405,203 @@ protected static function getMd5Version($path)
...
+									{
+										$includes[] = $found;
 
-									break;
+										break;
+									}
 								}
+								else
+								{
+									// Try to deals in the extension media folder
```
Comment (brianteeman) [inline libraries/src/HTML/HTMLHelper.php]: it was wrong before I know but lets get it right now. if i understand it correctly then it should be

"Try to deal with files in the extension's media folder"

File: libraries/src/HTML/HTMLHelper.php [hunk 3]
```diff
@@ -404,230 +405,203 @@ protected static function getMd5Version($path)
...
-						// Try to deal with system files in the media folder
-						else
-						{
-							$path = JPATH_ROOT . "/media/system/$folder/$file";
-
-							if (file_exists($path))
+							else
 							{
-								$includes[] = Uri::root(true) . "/media/system/$folder/$file" . static::getMd5Version($path);
+								// Try to deal with system files in the media folder
```
Comment (brianteeman) [inline libraries/src/HTML/HTMLHelper.php:574]: is this a duplicate of the code above?

Comment (dgrammatiko) [reply libraries/src/HTML/HTMLHelper.php:574]: yes, but the first part is in the `if` the second is in the `else`, I guess we can do a DRY-er approach here

File: libraries/src/HTML/HTMLHelper.php [hunk 4]
```diff
@@ -529,7 +529,7 @@ protected static function includeRelativeFiles($folder, $file, $relative, $detec
 								}
 								else
 								{
-									// Try to deals in the extension media folder
+									// Try to deal with files in the extension's media folde
```
Comment (brianteeman) [inline libraries/src/HTML/HTMLHelper.php]: folde => folder ;)

File: libraries/src/HTML/HTMLHelper.php [hunk 5]
```diff
@@ -1332,4 +1306,92 @@ public static function getJSObject(array $array = array())
...
+	protected static function convertToRelativePath($path)
+	{
+		$relativeFilePath = str_replace(JPATH_ROOT, '', $path);
+
+		// On windows devices we need to replace "\" with "/" otherwise some browsers will not load the asset
+		return str_replace(DIRECTORY_SEPARATOR, '/', $relativeFilePath);
+	}
+
+	/**
+	 * Method that takes two paths and checks if the files exist with diferent order
```
Comment (Quy) [inline libraries/src/HTML/HTMLHelper.php]: Spelling `diferent`

File: libraries/src/HTML/HTMLHelper.php [hunk 6]
```diff
@@ -1332,4 +1306,92 @@ public static function getJSObject(array $array = array())
...
+			$minifiedPath    = $path;
+			$nonMinifiedPath = str_replace('.min', '', $path, $position);
+
+			return self::checkFileOrder($nonMinifiedPath, $minifiedPath);
+		}
+
+		$minifiedPath = pathinfo($path, PATHINFO_DIRNAME) . '/' . pathinfo($path, PATHINFO_FILENAME) . '.min.' . $ext;
+
+		return self::checkFileOrder($minifiedPath, $path);
+
```
Comment (Quy) [inline libraries/src/HTML/HTMLHelper.php]: Remove blank line.

**Comment by brianteeman** [comment]
what if you want to load one of those specifically?

**Comment by dgrammatiko** [comment]
> what if you want to load one of those specifically?

Can you give me a practical example where this logic will be annoying?

**Comment by brianteeman** [comment]
Just thinking aloud. I am developing a site and I have not yet minified the files so I specifically reference the non minified file in my template. However because I do not have debug mode on I will see the old non-minified version

**Comment by dgrammatiko** [comment]
@brianteeman that's why Joomla 4 (thanks to @asika32764 ) provides a watch function for js and scss files for developers, so you'll never have to resolve a non issue 😏

**Comment by brianteeman** [comment]
Subject to that being documented my question is answered

**Comment by dgrammatiko** [comment]
### Documentation Changes Required

Well it's not the `detectBrowser` option that is changed here in a not B/C way. There is the drop of the `script.js / script-uncompressed.js` that is dropped as well in favour of only one way of defining compressible assets, like `script.min.js / script.js`. This change aligns joomla to the rest of the world about naming conventions of such assets and also reduces the processing power consumed having to check all the different conventions. Being different here is not helping us at all especially for the hustle we create for newcomers.

#### SO NO MORE script-uncompressed.js, DEAL WITH IT!

**Comment by mbabker** [comment]
It actually wasn't us being different for the sake of being different.  At one point that was the naming convention used for the MooTools files (by them, not us) and the rest of the API just followed suit on that convention until we started including files in core using the `.min.foo` convention and adding support for that instead of renaming everything we included.  So don't get angry with us on this one 😝

**Comment by dgrammatiko** [comment]
I'm not angry with nobody that put some code into the project. Well done, even if it was the wrong approach or not even delivering what was promised. Also I think that nobody will oppose fixing this now in J4, otherwise the project will continue currying the mootool bag for loooong long time. Now we have a chance to do it.

**Comment by astridx** [comment]
@dgt41 Is this PR ready for testing? 
If I assign this branch no css is loading in my case.
The path to the css files is not correct. Instead of
<tt><script src="/JOOMLA/joomla4/joomla-cms/media/mod_sampledata/js/sampledata-process.js"></script></tt>
the path is
<tt><script src="/media/mod_sampledata/js/sampledata-process.min.js"></script></tt>

**Comment by dgrammatiko** [comment]
@astridx that's interesting, probably this https://github.com/joomla/joomla-cms/pull/19744/files#diff-8f6485b51d191985c4d1e9fbbf1a743aR1364 doesn't work as expected. Are you on a windows machine?

**Comment by astridx** [comment]
No, I am not on Linux. I am on Ubuntu.
Edit: I meat I am not on windows :)

**Comment by dgrammatiko** [comment]
@astridx can you add this in you template somewhere:
```html
<?php echo JPATH_ROOT; ?>
```

**Comment by astridx** [comment]
I think JPATH_ROOT is OK. It is /var/www/html/JOOMLA/joomla4/joomla-cms as you can see on top of  the attached picture.
control panel test administration

**Comment by dgrammatiko** [comment]
@astridx can you post a screenshot of the head ?

**Comment by anuragteapot** [comment]
@dgt41 not able to load files. I download patch and tried to install. Installation not proceeded.

screenshot from 2018-03-15 21-04-41

**Comment by dgrammatiko** [comment]
@Anu1601CS is your installation on a subdirectory?

**Comment by brianteeman** [comment]
@dgt41 I think you should start to develop in a subdirectory :)

**Comment by anuragteapot** [comment]
> @Anu1601CS is your installation on a subdirectory?

No

**Comment by dgrammatiko** [comment]
@Anu1601CS ok can you share the head of the page that fails? what are the paths for these scripts?

**Comment by anuragteapot** [comment]
screenshot from 2018-03-15 22-09-58

**Comment by dgrammatiko** [comment]
@Anu1601CS the paths seem fine, but I still need the full src of the js/css there and in the image you shared after the ? everything gets hidden...

Also you do have the .min files in this installation? Right?

**Comment by dgrammatiko** [comment]
Forget it, I just saw that you are in a subdirectory...

**Comment by anuragteapot** [comment]
@dgt41 Oops, Now it works fine in the main directory as well in subdirectory :)

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on 77dd10258418b30a1b0a634c199d7edc50411c8d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19744">issues.joomla.org/tracker/joomla-cms/19744</a>.</sub>

**Comment by astridx** [comment]
I want to test this PR again. 

@dgt41 Sorry I missed the question for the head above. But now I send you my head :)

I am on branch <tt>4.0-dev</tt> in debug mode and everything is fine.

I fetch branch <tt>html/asset</tt> (git fetch origin pull/19744/head:html/assets) and change to branch <tt>html/asset</tt> (git checkout html/asset) and refesh the site I see no CSS. Do I make a mistake?

This is my head:


```
<!DOCTYPE html>
<html lang="en-gb" dir="ltr">
<head>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<meta name="theme-color" content="#1c3d5c">
	<meta name="generator" content="Joomla! - Open Source Content Management">
	<title>**Global Configuration** - tes - **Administration**</title>
	<link href="/JOOMLA/joomla4/joomla-cms/administrator/templates/atum/favicon.ico" rel="shortcut icon" type="image/vnd.microsoft.icon">

	<link href="/media/system/webcomponents/css/joomla-field-switcher.css?f40e1a10d455f4d4fa38109c590ae41a" rel="stylesheet" />
	<link href="/media/system/webcomponents/css/joomla-field-media.css?f40e1a10d455f4d4fa38109c590ae41a" rel="stylesheet" />
	<link href="/media/plg_system_debug/css/debug.css?f40e1a10d455f4d4fa38109c590ae41a" rel="stylesheet" />
	<link href="/administrator/templates/atum/css/bootstrap.css?f40e1a10d455f4d4fa38109c590ae41a" rel="stylesheet" />
	<link href="/administrator/templates/atum/css/font-awesome.css?f40e1a10d455f4d4fa38109c590ae41a" rel="stylesheet" />
	<link href="/administrator/templates/atum/css/template.css?f40e1a10d455f4d4fa38109c590ae41a" rel="stylesheet" />
	<link href="/media/vendor/joomla-custom-elements/css/joomla-alert.min.css?f40e1a10d455f4d4fa38109c590ae41a" rel="stylesheet" />
	<script type="application/json" class="joomla-script-options new">{
    "csrf.token": "645d62d47136387a418199d794e492c4",
    "system.paths": {
        "root": "\/JOOMLA\/joomla4\/joomla-cms",
        "rootFull": "http:\/\/localhost\/JOOMLA\/joomla4\/joomla-cms\/",
        "base": "\/JOOMLA\/joomla4\/joomla-cms\/administrator"
    },
    "joomla.jtext": {
        "JLIB_FORM_CONTAINS_INVALID_FIELDS": "**The form cannot be submitted as it's missing required data. <br> Please correct the marked fields and try again.**",
        "JLIB_FORM_FIELD_REQUIRED_VALUE": "**Field value cannot be empty**",
        "JLIB_FORM_FIELD_REQUIRED_CHECK": "**One of the options must be selected**",
        "JLIB_FORM_FIELD_INVALID_VALUE": "**This value is not valid**",
        "ERROR": "**Error**",
        "WARNING": "**Warning**",
        "NOTICE": "**Notice**",
        "MESSAGE": "**Message**",
        "JLIB_FORM_MEDIA_PREVIEW_EMPTY": "**No image selected.**",
        "JFIELD_PASSWORD_INDICATE_INCOMPLETE": "**Password doesn't meet site's requirements**",
        "JFIELD_PASSWORD_INDICATE_COMPLETE": "**Password accepted**",
        "JSHOW": "**Show**",
        "JHIDE": "**Hide**",
        "JLIB_JS_AJAX_ERROR_CONNECTION_ABORT": "**A connection abort has occurred while fetching the JSON data.**",
        "JLIB_JS_AJAX_ERROR_NO_CONTENT": "**No content was returned.**",
        "JLIB_JS_AJAX_ERROR_OTHER": "**An error has occurred while fetching the JSON data: HTTP %s status code.**",
        "JLIB_JS_AJAX_ERROR_PARSE": "**A parse error has occurred while processing the following JSON data:<br><code style=\"color:inherit;white-space:pre-wrap;padding:0;margin:0;border:0;background:inherit;\">%s<\/code>**",
        "JLIB_JS_AJAX_ERROR_TIMEOUT": "**A timeout has occurred while fetching the JSON data.**",
        "JLIB_HTML_PLEASE_MAKE_A_SELECTION_FROM_THE_LIST": "**Please first make a selection from the list.**"
    },
    "system.keepalive": {
        "interval": 840000,
        "uri": "\/JOOMLA\/joomla4\/joomla-cms\/administrator\/index.php?option=com_ajax&amp;format=json"
    },
    "webcomponents": {
        "joomla-switcher": "\/media\/system\/webcomponents\/js\/joomla-field-switcher.min.js?f40e1a10d455f4d4fa38109c590ae41a",
        "joomla-field-media": "\/media\/system\/webcomponents\/js\/joomla-field-media.js?f40e1a10d455f4d4fa38109c590ae41a",
        "joomla-field-send-mail": "\/media\/system\/webcomponents\/js\/joomla-field-send-test-mail.min.js?f40e1a10d455f4d4fa38109c590ae41a",
        "joomla-alert": "\/media\/vendor\/joomla-custom-elements\/js\/joomla-alert.min.js?f40e1a10d455f4d4fa38109c590ae41a"
    },
    "bootstrap.popover": {
        ".hasPopover": {
            "container": "body",
            "html": true,
            "trigger": "hover focus",
            "constraints": {
                "to": "scrollParent",
                "attachment": "together",
                "pin": true
            },
            "offset": "0 0"
        }
    },
    "keySave": {
        "task": "application.apply"
    }
}</script>

	<script src="/media/system/js/core.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/vendor/punycode/js/punycode.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/system/js/fields/validate.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/system/js/keepalive.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/vendor/jquery/js/jquery.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/legacy/js/jquery-noconflict.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/legacy/js/tabs-state.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/system/js/cms.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/vendor/bootstrap/js/bootstrap.bundle.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/legacy/js/bootstrap-init.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/legacy/js/helpsite.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/system/js/fields/passwordview.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/system/js/fields/permissions.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/plg_system_debug/js/debug.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/vendor/flying-focus-a11y/js/flying-focus.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/legacy/js/toolbar.js?f40e1a10d455f4d4fa38109c590ae41a"></script>
	<script src="/media/mod_menu/js/admin-menu.js?f40e1a10d455f4d4fa38109c590ae41a" defer></script>

</head>
```

instead of 


```
<!DOCTYPE html>
  | <html lang="en-gb" dir="ltr">
  | <head>
  | <meta charset="utf-8">
  | <meta name="viewport" content="width=device-width, initial-scale=1">
  | <meta name="theme-color" content="#1c3d5c">
  | <meta name="generator" content="Joomla! - Open Source Content Management">
  | <title>**Global Configuration** - tes - **Administration**</title>
  | <link href="/JOOMLA/joomla4/joomla-cms/administrator/templates/atum/favicon.ico" rel="shortcut icon" type="image/vnd.microsoft.icon">
  |  
  | <link href="/JOOMLA/joomla4/joomla-cms/media/system/webcomponents/css/joomla-field-switcher.css?8324ca156ad174f512939f1d000eb985" rel="stylesheet" />
  | <link href="/JOOMLA/joomla4/joomla-cms/media/system/webcomponents/css/joomla-field-media.css?8324ca156ad174f512939f1d000eb985" rel="stylesheet" />
  | <link href="/JOOMLA/joomla4/joomla-cms/media/plg_system_debug/css/debug.css?8324ca156ad174f512939f1d000eb985" rel="stylesheet" />
  | <link href="/JOOMLA/joomla4/joomla-cms/administrator/templates/atum/css/bootstrap.css?8324ca156ad174f512939f1d000eb985" rel="stylesheet" />
  | <link href="/JOOMLA/joomla4/joomla-cms/administrator/templates/atum/css/font-awesome.css?8324ca156ad174f512939f1d000eb985" rel="stylesheet" />
  | <link href="/JOOMLA/joomla4/joomla-cms/administrator/templates/atum/css/template.css?8324ca156ad174f512939f1d000eb985" rel="stylesheet" />
  | <link href="/JOOMLA/joomla4/joomla-cms/media/vendor/joomla-custom-elements/css/joomla-alert.min.css?8324ca156ad174f512939f1d000eb985" rel="stylesheet" />
  | <script type="application/json" class="joomla-script-options new">{
  | "csrf.token": "645d62d47136387a418199d794e492c4",
  | "system.paths": {
  | "root": "\/JOOMLA\/joomla4\/joomla-cms",
  | "rootFull": "http:\/\/localhost\/JOOMLA\/joomla4\/joomla-cms\/",
  | "base": "\/JOOMLA\/joomla4\/joomla-cms\/administrator"
  | },
  | "joomla.jtext": {
  | "JLIB_FORM_CONTAINS_INVALID_FIELDS": "**The form cannot be submitted as it's missing required data. <br> Please correct the marked fields and try again.**",
  | "JLIB_FORM_FIELD_REQUIRED_VALUE": "**Field value cannot be empty**",
  | "JLIB_FORM_FIELD_REQUIRED_CHECK": "**One of the options must be selected**",
  | "JLIB_FORM_FIELD_INVALID_VALUE": "**This value is not valid**",
  | "ERROR": "**Error**",
  | "WARNING": "**Warning**",
  | "NOTICE": "**Notice**",
  | "MESSAGE": "**Message**",
  | "JLIB_FORM_MEDIA_PREVIEW_EMPTY": "**No image selected.**",
  | "JFIELD_PASSWORD_INDICATE_INCOMPLETE": "**Password doesn't meet site's requirements**",
  | "JFIELD_PASSWORD_INDICATE_COMPLETE": "**Password accepted**",
  | "JSHOW": "**Show**",
  | "JHIDE": "**Hide**",
  | "JLIB_JS_AJAX_ERROR_CONNECTION_ABORT": "**A connection abort has occurred while fetching the JSON data.**",
  | "JLIB_JS_AJAX_ERROR_NO_CONTENT": "**No content was returned.**",
  | "JLIB_JS_AJAX_ERROR_OTHER": "**An error has occurred while fetching the JSON data: HTTP %s status code.**",
  | "JLIB_JS_AJAX_ERROR_PARSE": "**A parse error has occurred while processing the following JSON data:<br><code style=\"color:inherit;white-space:pre-wrap;padding:0;margin:0;border:0;background:inherit;\">%s<\/code>**",
  | "JLIB_JS_AJAX_ERROR_TIMEOUT": "**A timeout has occurred while fetching the JSON data.**",
  | "JLIB_HTML_PLEASE_MAKE_A_SELECTION_FROM_THE_LIST": "**Please first make a selection from the list.**"
  | },
  | "system.keepalive": {
  | "interval": 840000,
  | "uri": "\/JOOMLA\/joomla4\/joomla-cms\/administrator\/index.php?option=com_ajax&amp;format=json"
  | },
  | "webcomponents": {
  | "joomla-switcher": "\/JOOMLA\/joomla4\/joomla-cms\/media\/system\/webcomponents\/js\/joomla-field-switcher.min.js?8324ca156ad174f512939f1d000eb985",
  | "joomla-field-media": "\/JOOMLA\/joomla4\/joomla-cms\/media\/system\/webcomponents\/js\/joomla-field-media.js?8324ca156ad174f512939f1d000eb985",
  | "joomla-field-send-mail": "\/JOOMLA\/joomla4\/joomla-cms\/media\/system\/webcomponents\/js\/joomla-field-send-test-mail.min.js?8324ca156ad174f512939f1d000eb985",
  | "joomla-alert": "\/JOOMLA\/joomla4\/joomla-cms\/media\/vendor\/joomla-custom-elements\/js\/joomla-alert.min.js?8324ca156ad174f512939f1d000eb985"
  | },
  | "bootstrap.popover": {
  | ".hasPopover": {
  | "container": "body",
  | "html": true,
  | "trigger": "hover focus",
  | "constraints": {
  | "to": "scrollParent",
  | "attachment": "together",
  | "pin": true
  | },
  | "offset": "0 0"
  | }
  | },
  | "keySave": {
  | "task": "application.apply"
  | }
  | }</script>
  |  
  | <script src="/JOOMLA/joomla4/joomla-cms/media/system/js/core.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/vendor/punycode/js/punycode.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/system/js/fields/validate.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/system/js/keepalive.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/vendor/jquery/js/jquery.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/legacy/js/jquery-noconflict.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/legacy/js/tabs-state.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/system/js/cms.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/vendor/bootstrap/js/bootstrap.bundle.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/legacy/js/bootstrap-init.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/legacy/js/helpsite.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/system/js/fields/passwordview.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/system/js/fields/permissions.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/plg_system_debug/js/debug.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/vendor/flying-focus-a11y/js/flying-focus.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/legacy/js/toolbar.js?8324ca156ad174f512939f1d000eb985"></script>
  | <script src="/JOOMLA/joomla4/joomla-cms/media/mod_menu/js/admin-menu.js?8324ca156ad174f512939f1d000eb985" defer></script>
  |  
  | </head>
 ```

**Comment by dgrammatiko** [comment]
@astridx can you edit line 1367 of HTMLHelper.php to:
```php
$relativeFilePath = Uri::root(true) . str_replace(JPATH_ROOT, '', $path);
```
Does this solve the problem?

**Comment by astridx** [comment]
Yes, i changed 

`$relativeFilePath = str_replace(JPATH_ROOT, '', $path);
`
to
`$relativeFilePath = Uri::root(true) . str_replace(JPATH_ROOT, '', $path);`
and now CSS is loaded.

**Comment by astridx** [comment]
I have tested this item :white_check_mark: successfully on d3503455e36ba24107aa94100f4f320a7c22ffe3<br /><br />That is what I tested:


I have activated debug mode.
I change to your branch and I see that core.js is loaded.
I renamed core.js.
I refresh page and I see that core.min.js is loaded.
I deactivated debug mode and I see that core.min.js is loaded
I revert renaming of core.js and renamed core.min.js and I see, that core.js is leaded.

Without applying this patch
In no-debug-mode i need a core.min.js. If there is only a core.js no core styles are loaded.
In debug-mode core.js is loaded. If there is no core.js the minified version is loaded.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19744">issues.joomla.org/tracker/joomla-cms/19744</a>.</sub>

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on d3503455e36ba24107aa94100f4f320a7c22ffe3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19744">issues.joomla.org/tracker/joomla-cms/19744</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by dgrammatiko** [comment]
@mbabker @laoneo @wilsonge is there some interest for this one or shall I close it?

**Comment by mbabker** [comment]
IMO yes get it merged.

File: libraries/src/HTML/HTMLHelper.php [hunk 7]
```diff
@@ -16,10 +16,7 @@
 use Joomla\CMS\Log\Log;
 use Joomla\CMS\Uri\Uri;
 use Joomla\Utilities\ArrayHelper;
-
-\JLoader::import('joomla.environment.browser');
-\JLoader::import('joomla.filesystem.file');
-\JLoader::import('joomla.filesystem.path');
+use Joomla\Filesystem\File;
```
Comment (laoneo) [inline libraries/src/HTML/HTMLHelper.php]: This should be `use Joomla\CMS\Filesystem\File;`.

**Comment by laoneo** [comment]
No issues with the proposal. What should we do with the tests? Also let's fix the conflicts.

**Comment by dgrammatiko** [comment]
Ok, this is synced once more

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 9a1840e05e900b89efed1f78a9f22791346da3c1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19744">issues.joomla.org/tracker/joomla-cms/19744</a>.</sub>

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 9a1840e05e900b89efed1f78a9f22791346da3c1<br /><br />Like this change!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19744">issues.joomla.org/tracker/joomla-cms/19744</a>.</sub>
